src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,272 @@
+            {
+                return false;
+            }
+
+            return !(session.Properties.TryGetProperty(CompletionSource.ExcludedCommitCharacters, out ImmutableArray<char> excludedCommitCharacter)
+                && excludedCommitCharacter.Contains(typedChar));
+        }
+
+        public AsyncCompletionData.CommitResult TryCommit(
+            IAsyncCompletionSession session,
+            ITextBuffer subjectBuffer,
+            VSCompletionItem item,
+            char typeChar,
+            CancellationToken cancellationToken)
+        {
+            // We can make changes to buffers. We would like to be sure nobody can change them at the same time.
+            AssertIsForeground();
+
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return CommitResultUnhandled;
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+            if (completionService == null)
+            {
+                return CommitResultUnhandled;
+            }
+
+            if (!item.Properties.TryGetProperty(CompletionSource.RoslynItem, out RoslynCompletionItem roslynItem))
+            {
+                // Roslyn should not be called if the item committing was not provided by Roslyn.
+                return CommitResultUnhandled;
+            }
+
+            var filterText = session.ApplicableToSpan.GetText(session.ApplicableToSpan.TextBuffer.CurrentSnapshot) + typeChar;
+            if (Helpers.IsFilterCharacter(roslynItem, typeChar, filterText))
+            { 
+                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+            }
+
+            var serviceRules = completionService.GetRules();
+
+            // We can be called before for ShouldCommitCompletion. However, that call does not provide rules applied for the completion item.
+            // Now we check for the commit charcter in the context of Rules that could change the list of commit characters.
+
+            // Tab, Enter and Null (call invoke commit) are always commit characters. 
+            if (typeChar != '\t' && typeChar != '\n' && typeChar != '\0' && !IsCommitCharacter(serviceRules, roslynItem, typeChar, filterText))
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+            }
+
+            if (!session.Properties.TryGetProperty(CompletionSource.TriggerSnapshot, out ITextSnapshot triggerSnapshot))
+            {
+                // Need the trigger snapshot to calculate the span when the commit changes to be applied.
+                // It should be inserted into a property bag within GetCompletionContextAsync for each item created by Roslyn.
+                // If not found here, Roslyn should not make a commit.
```

Comment:
How would we get here in the first place? You already filtered non-Roslyn items earlier?

Reply:
1. We can miss to provide triggerSnapshot by mistake.

2. Let us assume there is another provider wants Roslyn to perform a commit and provided CompletionSource.RoslynItem above. We can proceed but we also need triggerSnapshot in the case.


---
In reply to: [236497307](https://github.com/dotnet/roslyn/pull/29016#discussion_r236497307) [](ancestors = 236497307)